### ISSUE TITLE
Project Panels interface confusing and does not work as expected.  Needs overhaul... I am losing project files consistently...

### ISSUE THREAD

**Issue by richlysakowski** [notepad-plus-plus/notepad-plus-plus#13131]
First of all I want to say that I LOVE Notepad++.  I use it all day, every day... for text editing, note taking, fixing code, saving code, cleaning data, saving valuable snippets, organizing ideas in hierarchical format (yml), reformatting data (JSON, HTML, etc.), inspecting CSVs, and helping me keep track of things on my computer and in my daily life as a computer user, power user, and programmer.  

I use Notepad++ more than almost any other tool on every system I use routinely.  (If NPP ran on my Android cell phone I would use it there).   

And I want to use it **_MORE_**, **for projects**.  However, the Project Panels interface needs  significant work to be REALLY useful.  I cannot easily find and open the location of the current Project Panel.  Its location is available (if it was saved) by hovering over the icon at the top of the panel, but I cannot right-click and open that location, I have to hover-and-remember, and then navigate there to access it directly or move to a better location.  I am losing Project Panel XML files consistently...  

There needs to be a proper Project Browser.  
The ability to rename a Project Panel and have the new name appear on the tab at the bottom of the NPP window, just above the Status Bar.  

Project Panels need a "Properties Panel" or dialog box that can be opened to see the current properties for each project, including: 

- Current Project Panel file name
- Project Panel XML file location, 
- Other data pertinent to the group of files, perhaps annotation made by the user.
- When the Project Panel was created (perhaps recent revision history?)
- Who created the file.  
- Perhaps a button to move (i.e., "Rename") the XML file to another location
- Names and files currently referenced in the Project Panel XML file 

**Project Register:** 
A centralized register of all Project Panels, so one can keep track of the names and locations of files in a project panel group.   One central place to find all my Project Panel XML files, whether the XML file is local or elsewhere.  

### Description of the Issue
See above. 

### Steps to Reproduce the Issue
See below...

### Expected Behavior
I rename project panels and their new name will appear in the lower-left tabbed bar at the bottom of NPP main window.  

I right click on a project panel tab and it shows me its full path, and gives me an option to open it.  

I look in my project panel registry and I can see all kinds of important information about all my project panels, especially their location.  

### Actual Behavior

Project Panel tabs only ever say "Project Panel 1", "Project Panel 2", "Project Panel 3", 

When I right click on a project panel tab, it does not show me its full path.  It does not give me an option to open it.   (see screenshot below). 

There is no central registry for project panel files; they are scattered all over the place with no easy way to locate or manage them.  

The instructions for how to use the Project Panel feature is hard to understand and confusingly mixed together with "Workspace as Project", 

The best confusing examples are 

"Save As = save the active Workspace in a new Windows filesystem location."  
    **_To be consistent with the rest of NPP, this should be "Rename", not "Save As..."_**

"Save As Copy = save the active Workspace in a new Windows filesystem location, keeping the old Workspace file as well."
    **_To be consistent with the rest of NPP, this should be just "Save As..."_**

Personally I think Workspace XML files should have their own file extension, so they are easier to distinguish from other XML files, and can be searched for by type.  An extension like ".npp_ws" or ".npp_workspace", e.g., my_project_workspace.npp_ws 

**_DO NOT USE ".ws" for a workspace file extension!!_**   See What is a WS file?  (Windows Server Script, WonderSwan, 1-2-3 WordSearch, WhiteSpace Source Code).  Executable script for Windows that can incorporate JScript and VBScript routines and may include XML elements; runs when opened if Windows Scripting Host (WSH) 2.0 or later is installed.

![image](https://user-images.githubusercontent.com/9345711/218942556-9a1b51c4-ce37-4a56-80b0-747b02eef3b3.png)

**Comment by alankilborn** [comment]
I tend to agree with the overall sentiment:  "Projects" are not handled all that well in N++.  That being said I have some general comments:

----------------------

>The instructions for how to use the Project Panel feature is hard to understand and confusingly mixed together with "Workspace as Project"

This is a user manual complaint for this general area: https://npp-user-manual.org/docs/session/
@pryrt might be interested in knowing your feelings about user manual deficiencies.

----------------------

>I think Workspace XML files should have their own file extension

I use `.workspace` myself, but maybe it is reasonable for N++ to just state "foo is the extension for workspace files" instead of making it configurable (without a default) at _Prefs_ > _MISC_

---------------------------------

>There is no central registry for project panel files; they are scattered all over the place with no easy way to locate or manage them.

Not sure I follow this...

----------------------------------

>I am losing Project Panel XML files consistently...

Truly losing them? Or just not remembering where you put a workspace file

I have a a folder called "Workspaces" and I put, well, workspace files there.  At least this way I know where to find them all.

----------------------------------

Since you are interested in projects/workspaces, note that I have created some other issues regarding them recently.

**Comment by pryrt** [comment]
> The instructions for how to use the Project Panel feature is hard to understand and confusingly mixed together with "Workspace as Project",

They've each got their own section on the page.  And the Project Panel has the most description of all the sessions, workspaces, and project features.  And the Project Panel has an entire paragraph on the differences between it and "Folder as Workspace". I'm sorry all that detail confuses you.

> The best confusing examples are
> 
> "Save As = save the active Workspace in a new Windows filesystem location." **_To be consistent with the rest of NPP, this should be "Rename", not "Save As..."_**

The **File > Save As...** entry in the main NPP interface would beg to differ.  The Project **Save As** feature and the NPP **File > Save As** both do the same thing.  If you are editing a file called `old.txt` in Notepad++, and do a **Save As** to `new.txt`, then `new.txt` will be saved and you will continue editing `new.txt`, but `old.txt` will still exist.  Similarly, if you have a project called `project1.wkspc` and do a **Save As** to `project2.wkspc`, the `project2.wkspc` will be saved and that will be the active project file, but the old file `project1.wkspc` will still exist.  This is _not_ a rename function, this is a Save As function, as both the GUI and the user manual describe.  Apparently, my description does not make this clear to you.

If there were a "Rename" function for projects, the effect would be to delete the file `project1.wkspc` and replace it with `project2.wkspc` so that `project1.wkspc` no longer existed.  But that's _not_ what Project > **Save As** does.

> "Save As Copy = save the active Workspace in a new Windows filesystem location, keeping the old Workspace file as well." **_To be consistent with the rest of NPP, this should be just "Save As..."_**

The **File > Save a Copy As** entry in the main NPP interface would beg to differ.  The Project **Save a Copy As** feature and the NPP **File > Save a Copy As** feature both do the same thing.  If you are editing a file called `old.txt` in Notepad++, and do a **Save a Copy As** to `new.txt`, then `new.txt` will be saved, but you will continue editing `old.txt`, but `old.txt` and `new.txt` will both co-exist.  Similarly, if you have a project called `project1.wkspc` and do a **Save a Copy As** to `project2.wkspc`, the `project2.wkspc` will be saved, but `project1.wkspc`will still be the active project file, and both files will co-exist.  This is _not_ a simple Save As function, because the copy gets saved but you continue using the old version rather than switching to the new version.  This is a "Save a Copy As" function, as both the GUI and the user manual describe.  Apparently, my description does not make this clear to you.

I had assumed that users knew the traditional differences between **Save As** and **Save a Copy As**, because both of those are standard features of editors, and both exist in NPP already, so I gave a rather brief description of their purpose.  Apparently, that assumption was foolish of me, and I will clarify those: I will clarify with examples which files still exist, and which file is active (similar to my paragraph's here).  And I will change the text in the User Manual from "Save As Copy" to "Save a Copy As", to match the GUI more closely.

**Comment by pryrt** [comment]
That was weird. I went to re-run the experiment, and now it seems that **Workspace > Save a Copy As** is actually behaving identically to **Workspace > Save As**, rather than like the **Save a Copy As** that I described above (and have now crossed out).

@alankilborn , maybe I've just confused myself.  Could you try the sequences that I described earlier, and see whether SaCA does what I originally said or whether it replicates SaveAs?  Because if it really does exactly what SaveAs does, then it's a bug in the implementation.

**Comment by alankilborn** [comment]
@pryrt If you can shorten your "Conclusion" text and use it as a title, and add _Debug Info_, I think you have the good-makins of a "New Issue".  :-)

I must admit that in real world use, I have never used SaCA on a workspace.

**Comment by pryrt** [comment]
Neither had I.  Which is why I was surprised and confused (and initially mis-described the bug)

Created new issue for that specific bug.

**Comment by pryrt** [comment]
@richlysakowski ,

Do you think these descriptions better explain **Save As** vs **Save a Copy As**?

![image](https://user-images.githubusercontent.com/17455758/219106780-71795a56-afd6-45ce-87a3-d68877cb3075.png)

**Comment by pryrt** [comment]
> I cannot easily find and open the location of the current Project Panel. Its location is available (if it was saved) by hovering over the icon at the top of the panel, but I cannot right-click and open that location, I have to hover-and-remember, and then navigate there to access it directly

Yes, "Open Location in Explorer" and "Open Location in cmd" and "Copy workspace-file location to clipboard" would be great actions to add to the Workspace right-click context menu and **Workspace** toolbar menu.

**Comment by alankilborn** [comment]
>But if you were giving a user an option of where to default Workspace files...

I really wasn't getting that far into it.  N++'s workspace/projects handling is so infantile (sorry), that for me, a central location that I've set up for all of the ones I used works best for me.  Surely the most flexible (and settings-intense) capability would be awesome, but remember the mantra of the Notepad++ author:  "too many options leads to no-implementation" or something like that (I think the actual thing he said one time was "too many options equal no option", but that's how I interpret that.  Truly, @pryrt I think any and all of your proposed ideas make sense.

**Comment by pryrt** [comment]
> _VS **Recent Projects and Solutions** menu entry_

Ooh, that's the best idea: **Recent Workspaces and Sessions**.  As long as it's defined so that as soon as you **Open** or **Save As** the workspace, it will be added to the list, rather than only when a workspace is closed.  And that follows the Philosophy of Least Options.

**Comment by alankilborn** [comment]
>Ooh, that's the best idea: Recent Workspaces and Sessions

Perhaps I'll break that one off into its own issue as a Feature Request, but I'll take the liberty to suggest it as _Recent Sessions and Workspaces_ as I think sessions are probably more popular among N++ users than workspaces.

>For the entries in Recent Workspaces and Sessions, I would think each would indicate (either with icon or a prefix) that it's either a Workspace or Session file.

Or:

* idea 1: _File_ > _Recent Sessions_ > ... **_and_** _File_ > _Recent Workspaces_ > ...

* idea 2: _File_ > _Recent Sessions and Workspaces_, and then sessions listed first (upper part) and a separator and then workspaces listed (lower part)

**Comment by pryrt** [comment]
I think idea 2 is best.

**Comment by richlysakowski** [comment]
Your suggestion is excellent.  The next step should be something like you
suggest.

I am not a good C or C++ coder (Python & SQL, but not C++) or I would try
to add it myself.  I can try to hack it out myself.

If someone wants to jump in and guide or help me I would appreciate it.

Reuse and Repurpose:

Perhaps we could steal / reuse code from other parts of NPP to get it
working.

Here's a few next steps I see...

The most "Recently Used Project..." menu could be copied from "Recently
Used File..." menu.

The history stack or queue data structure to save, retrieve and display the
last N projects could be reused from the Recently Used File..." menu also.

One design question is whether the Project History menu should be a
right-click menu function that is local and bound to the current Project
Panel?  Or should it be a function on main application File Menu?  Visual
Studio has it as a global menu, and does not have the same idea of Project
Panels.

The multiple Project Panels tabbed interface is a nice idea because people
usually have more than different workspace and project contexts.

What do people think about this?  Should multiple Project Panels tabbed
interface  be kept, or reduced to one Project Panel at a time?

#######################

Instant, Easy Improvement for the "Save" workspace misnomer:

This improvement should be trivial, to rename the Project Panel "Save"
misnomer to call it "Rename", because that is what it really does.

Rich

**Comment by richlysakowski** [comment]
Yes, great suggestion... Sessions and Workspaces are very different
concepts and should be VISUALLY distinguished.

SESSIONS save the file working set when Windows crashes.  (This is one of
the most important features of NPP, because it prevents data loss.  Windows
NotePad loses everything that was not consciously saved before a Windows
crash.)

WORKSPACES are user-defined and named collections of Projects.

##############

I just recently found (for the first time) the NPP backup files inside the
hidden NPP session backup folder.  I thought to myself, so THIS is where
all my sessions are centralized!  Perhaps I could use this to recover my
lost and scattered Project Panels?

Rich

**Comment by richlysakowski** [comment]
great idea!

**Comment by alankilborn** [comment]
I don't know that more than 1 project panel available simultaneously is valuable (currently N++ has 3).  I tend to have a session dedicated to a project -- if this is a common way to work then does a "session" and a "workspace" need to be two separate things in N++?

Truly I ask myself, Why did I even set this up?  And the answer is, I needed a way to search certain .c and .h files and not others that are intermingled, across several directories (so that the _Find in Files_ _Directory_ and _Filters_ parameters just don't cut it for what I need to search -- I can define a "project" and then search only files in the project.  Sure, I could search ALL .c and .h files, and then mentally ignore certain ones, and I WAS doing this for a while, but it got brain-intensive to do this mental filtering.

So it seems like sessions came first, and sessions do indeed make a lot of sense.  And then workspace/projects were "bolted on" and it got kind of kludgey.

Can it all be sorted out and streamlined into something really useful and sensical?  Unknown, but perhaps someone should attempt it.  And no, that won't be me.  :-)

**Comment by alankilborn** [comment]
BTW my last comment brings up another point:  Why can't I just tell N++ to search all of the current project's files in an easier fashion than having to move to the _Filters_ box and set it to `*.*`?  I suppose the flexibility to specify something there that isn't `*.*` and have it only search a subset of the project's files has some sense, though...

**Comment by richlysakowski** [comment]
The important concept here is first a central registry, with pointers to
the Workspace and Project files.  The registry and storage do not have to
be the same place.  The Registry should be a centralized index, and storage
could be distributed.

If storage is NOT centralized (like conda does for all virtual environments
in the Python ecosystem), then when people move workspace or project files,
the Workspace/Project Registry can get out of sync.   There would then need
to be a file system scan to find the moved Workspace or Project XML file
and update the registry.

Personally I prefer to have both indexing and physical storage centralized,
because it is faster to update and manage things.  However, this requires
that the centralized storage be large enough to accommodate all projects
and Workspaces.  Data can be distributed more easily than code (yeah, I
know "code is data")

I am sure I am not the only developer with a bajillion files, a hoarding
instinct  and a poor memory.  It is more work for me when things are
distributed / scattered.

How to best (simply) organize Workspaces, Projects, Files for
scatter-brained developers?

Can we design the tools to help organize the disorganized?  Perhaps NO for
the hopelessly disorganized, but YES for the mildly disorganized {8^)?

Incremental improvements are the way to go.  After all, this is a volunteer
effort.

On the other hand, maybe we could crowd-source fund the bigger
improvements?   Kick-Starter project anyone?

Rich


On Wed, Feb 15, 2023, 12:58 PM PeterCJ ***@***.***> wrote:

> *I am losing project files consistently*
> ...
> There is no central registry for project panel files; they are scattered
> all over the place with no easy way to locate or manage them.
>
> If I've understood correctly, I think you mean that you would like to have
> one directory where you store all your Project Panel's Workspace files, so
> that you can just go to that one central location.
>
> From my perspective, the reason that Notepad++ doesn't default to a
> "central workspace-file location" is because a workspace project file is so
> closely tied with the individual project you are editing. In a tool like
> Visual Studio, the .vcxproj file stays near the source code for that
> particular project, it doesn't go off in some "here are all my project
> files" directory. So the developers decided to default the workspace-file
> location to be with the files in the project.
>
> But don't get me wrong, I can easily understand the "I don't remember what
> folder my source code is in... I want to be able to open the Project
> Workspace from a central location" use-case.
>
Or similarly, the use-case of "my project's files are spread throughout

**Comment by pryrt** [comment]
@richlysakowski ,

>SESSIONS save the file working set when Windows crashes. 

The main/default Notepad++ session does that.  The Load Session / Save Session is separate, and is used by a significant portion of the Notepad++ users as a simple version of a "project".  That's the whole reason the Session Manager plugin exists: because people wanted something that could save a group of related files, and be able to open them all on demand.  The fact that Project Workspaces are better suited for that task doesn't change the fact that it's how the Load/Save Session feature and the Session Manager plugin _are_ used.

<hr>

> Incremental improvements are the way to go.  

I'm glad you said that (though it almost got buried in the other content of your replies, and so this reply was almost very different).  Because with some of the other stuff you were still saying, I was worried that you were going to get too much feature-creep, and then run into disappointment when the entire PR was rejected -- and that would have the side effect of the developer being soured on _any_ solution to the issue(s) involved.

And since you said, "I am not a good C or C++ coder ... or I would try to add it myself.  I can try to hack it out myself."  That's a great attitude.  But if the developer doesn't like the _specific_ implementation, it again runs the risk of grounding any effort to get this new feature.

<hr>

To bring us back to a summary of where we are for the various sub-issues you originally brought up:

1. "losing project files".  I think the best chance we have is of getting **Recent Session and Workspace** menu.  And @alankilborn [already indicated](#issuecomment-1431830938) that it should probably be made a separate issue, and my interpretation was that he'd be submitting that request.

2. Documentation: I've already made improvements, to help compare/contrast SaveAs and SaveACopyAs, which will probably be in the next version of the user manual website.

3. **Save A Copy As** bug: I already put in a bug report #13135 for that one.

4. Adding "Open Location in Explorer" and "Open Location in cmd" and "Copy location to clipboard" entries to the Workspace right-click context menu and Workspace toolbar menu would be a reasonable incremental solution.  But as a request buried in this multi-faceted issue, that's going to be lost and will be forgotten (I had nearly forgotten it already in the last few hours).  If you'd like to see that feature added, I suggest making a new, focused Feature Request issue that just requests those three menu entries.

I think that covers the major issues/suggestions you had.
